packages/@aws-cdk/core-synthesizer/lib/synthesizer.ts

```diff
@@ -0,0 +1,79 @@
+import { App, AssetManifestBuilder, DockerImageAssetLocation, DockerImageAssetSource, FileAssetLocation, FileAssetSource, IBoundStackSynthesizer, IReusableStackSynthesizer, ISynthesisSession, Stack, StackSynthesizer } from '@aws-cdk/core';
+import { assertBound } from '@aws-cdk/core/lib/stack-synthesizers/_shared';
+import { IStagingStack, StagingStack } from './staging-stack';
+
+/**
+ * New stack synthesizer properties
+ */
+export interface NewStackSynthesizerProps {
+  /**
+   * Bring a custom staging stack into the app.
+   *
+   * @default - default staging stack
+   */
+  readonly stagingStack?: IStagingStack;
+}
+
+/**
+ * New Stack Synthesizer
+ */
+export class NewStackSynthesizer extends StackSynthesizer implements IReusableStackSynthesizer, IBoundStackSynthesizer {
+  private qualifier?: string;
+  private stagingStack: IStagingStack;
+  private assetManifest = new AssetManifestBuilder();
+
+  constructor(props: NewStackSynthesizerProps = {}) {
+    super();
+
+    // TODO: ensure no tokens
+
+    // TODO: revisit scope
+    this.stagingStack = props.stagingStack ?? new StagingStack(new App(), 'StagingStack', {});
+  }
+
+  /**
+   * // TODO
+   */
+  public reusableBind(stack: Stack): IBoundStackSynthesizer {
+    // Create a copy of the current object and bind that
+    const copy = Object.create(this);
+    copy.bind(stack);
+    return copy;
```

Comment:
This is a hacky implementation for backwards compatibility and minimal code changes to the old code base. The nicer way is two have two separate classes:

`NewStackSynthesizer` and `BoundNewStackSynthesizer`:

```ts
export class NewStackSynthesizer {
  reusableBind(stack) {
    return new BoundNewStackSynthesizer(stack, this.configurationPropsOrWhatever);
  }

  bind(stack) {
    throw new Error('This is a legacy API, call reusableBind instead'); // Won't be hit anyway
  }

  synthesize(stack) {
    throw new Error('This is a legacy API, call reusableBind instead'); // Won't be hit anyway
  }
}

// Doesn't even need to be exported
class BoundNewStackSynthesizer {
}
```